src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+  private HttpAsyncClientBuilder customizeHttpClientConfig(HttpAsyncClientBuilder builder) {
+    builder.setMaxConnPerRoute(config.maxInFlightRequests());
+    configureAuthentication(builder);
+
+    if (config.secured()) {
+      log.info("Using secured connection to {}.", config.connectionUrls());
+      configureSslContext(builder);
+    } else {
+      log.info("Using unsecured connection to {}.", config.connectionUrls());
+    }
+    return builder;
+  }
 
   /**
-   * Executes a bulk action.
+   * Customizes each request according to configurations.
    *
-   * @param bulk the bulk request
-   * @return the bulk response
-   * @throws IOException if the client cannot execute the request
+   * @param builder the RequestConfigBuilder
+   * @return the builder
    */
-  BulkResponse executeBulk(BulkRequest bulk) throws IOException;
+  private RequestConfig.Builder customizeRequestConfig(RequestConfig.Builder builder) {
+    return builder
+        .setContentCompressionEnabled(config.compression())
+        .setConnectTimeout(config.connectionTimeoutMs())
+        .setConnectionRequestTimeout(config.readTimeoutMs());
+  }
+
+  /**
+   * Gets the mapping for an index.
+   *
+   * @param index the index to fetch the mapping for
+   * @return the MappingMetaData for the index
+   */
+  private MappingMetaData getMapping(String index) {
+    GetMappingsRequest request = new GetMappingsRequest().indices(index);
+    GetMappingsResponse response = callWithRetries(
+        "get mapping for index " + index,
+        () -> client.indices().getMapping(request, RequestOptions.DEFAULT)
+    );
+    return response.mappings().get(index);
+  }
 
   /**
-   * Executes a search.
+   * Processes a response from a BulkAction. Successful responses are ignored. Failed responses are
+   * reported to the DLQ and handled according to configuration (ignore or fail). Version conflicts
+   * are ignored.
    *
-   * @param query the search query
-   * @param index the index to search
-   * @param type the type to search
-   * @return the search result
-   * @throws IOException if the client cannot execute the request
+   * @param response the response to process
    */
-  JsonObject search(String query, String index, String type) throws IOException;
+  private void handleResponse(BulkItemResponse response) {
+    if (response.isFailed()) {
+      for (String error : MALFORMED_DOC_ERRORS) {
+        if (response.getFailureMessage().contains(error)) {
+          handleMalformedDocResponse(response);
+          reportBadRecord(response);
+          return;
+        }
+      }
+
+      if (response.getFailureMessage().contains(VERSION_CONFLICT_EXCEPTION)) {
+        log.warn(
+            "Ignoring version conflict for operation {} on document '{}' version {} in index '{}'.",
+            response.getOpType(),
+            response.getId(),
+            response.getVersion(),
+            response.getIndex()
+        );
+
+        reportBadRecord(response);
```

Comment:
The method's Javadoc and the `warn` message says we are ignoring version conflict,s but it seems like we are still reporting this record. Is that intentional?

Reply:
reporting the record is the same as ignoring it (reporting a bad record does not fail anything), but I figured the user should be know which records had version conflicts in case they want to remediate manually somehow